tests/unittests/GraphOptzTest.cpp

```diff
@@ -2882,3 +2894,92 @@ TEST_F(GraphOptz, nopRelu) {
+  TH.clear(value);
+}
+
+TEST_F(GraphOptz, constantFoldSingleNode) {
+  auto *const1 = mod_.createConstant(ElemKind::FloatTy, {2, 2}, "const1");
+  auto *const2 = mod_.createConstant(ElemKind::FloatTy, {2, 2}, "const2");
+  auto *ph1 = mod_.createPlaceholder(ElemKind::FloatTy, {2, 2}, "input1",
+                                     /* isTrainable */ false);
+  setConstValue(const1, 1.0f);
+  setConstValue(const2, 2.0f);
+  auto *splat2 = F_->createSplat(
+      "splat2", mod_.uniqueType(ElemKind::FloatTy, {2, 2}), 2.0f);
+  auto *splat3 = F_->createSplat(
+      "splat3", mod_.uniqueType(ElemKind::FloatTy, {2, 2}), 3.0f);
+
+  auto *add1 = F_->createAdd("add", const1, const2);
+  auto *mul1 = F_->createMul("mul1", add1, splat2);
+  auto *mul2 = F_->createMul("mul2", mul1, splat3);
+  auto *SN1 = F_->createSave("save", mul2);
+  auto *add3 = F_->createAdd("add", const1, ph1);
+  auto *SN2 = F_->createSave("save", add3);
+
+  // Perform constant folding for a specific node.
+  std::vector<Constant *> constResults =
+      constantFold(SN1->getInput().getNode());
+
+  EXPECT_EQ(constResults.size(), 1);
+  SN1->getInput().replaceAllUsesOfWith(constResults[0]);
+  // Second save should be unaffected.
+  EXPECT_FALSE(llvm::isa<Constant>(SN2->getInput()));
+  // First save should have been constant folded.
+  EXPECT_TRUE(llvm::isa<Constant>(SN1->getInput()));
+  Constant *C = llvm::dyn_cast<Constant>(SN1->getInput());
+  auto CH = C->getHandle();
+  // The expected result should be: (((1+2) * 2 * 3) = 18
+  EXPECT_EQ(CH.at({0, 0}), 18.0f);
+  EXPECT_EQ(CH.at({0, 1}), 18.0f);
+  EXPECT_EQ(CH.at({1, 0}), 18.0f);
+  EXPECT_EQ(CH.at({1, 1}), 18.0f);
+}
+
+TEST_F(GraphOptz, constantFoldWholeFunction) {
+  auto *const1 = mod_.createConstant(ElemKind::FloatTy, {2, 2}, "const1");
+  auto *const2 = mod_.createConstant(ElemKind::FloatTy, {2, 2}, "const2");
+  auto *const3 = mod_.createConstant(ElemKind::FloatTy, {2, 2}, "const3");
+  auto *const4 = mod_.createConstant(ElemKind::FloatTy, {2, 2}, "const4");
+  auto *ph1 = mod_.createPlaceholder(ElemKind::FloatTy, {2, 2}, "input1",
+                                     /* isTrainable */ false);
+  setConstValue(const1, 1.0f);
+  setConstValue(const2, 2.0f);
+  setConstValue(const3, 3.0f);
+  setConstValue(const4, 4.0f);
+  auto *splat2 = F_->createSplat(
+      "splat2", mod_.uniqueType(ElemKind::FloatTy, {2, 2}), 2.0f);
+  auto *splat3 = F_->createSplat(
+      "splat2", mod_.uniqueType(ElemKind::FloatTy, {2, 2}), 3.0f);
+  auto *splat4 = F_->createSplat(
+      "splat2", mod_.uniqueType(ElemKind::FloatTy, {2, 2}), 4.0f);
+
+  auto *add1 = F_->createAdd("add", const1, const2);
+  auto *mul1 = F_->createMul("mul1", add1, splat2);
+  auto *mul2 = F_->createMul("mul2", mul1, splat3);
+  auto *sub = F_->createSub("sub", mul2, const3);
+  auto *add2 = F_->createAdd("add2", sub, const4);
+  auto *mul3 = F_->createMul("mul3", add2, splat4);
+  auto *SN1 = F_->createSave("save", mul3);
```

Comment:
Can you add a TopK or something with multiple results that would be the final Node in the Constant tree, so we test correctness of ConstantProp with more than 1 result?

Reply:
Fixed.